contracts/NewBlockRelay.sol

```diff
@@ -0,0 +1,380 @@
+      candidates.push(vote);
+      // Mapping the vote into its hashes
+      voteInfo[vote].voteHashes.blockHash = _blockHash;
+      voteInfo[vote].voteHashes.drMerkleRoot = _drMerkleRoot;
+      voteInfo[vote].voteHashes.tallyMerkleRoot = _tallyMerkleRoot;
+      voteInfo[vote].voteHashes.previousVote = _previousVote;
+      voteInfo[vote].voteHashes.epoch = _epoch;
+    }
+
+    // Sum one vote
+    voteInfo[vote].numberOfVotes += 1;
+    // Update the block with more votes
+    if (vote != winnerVote) {
+      // Set as new winner if it has more votes
+      if (voteInfo[vote].numberOfVotes > voteInfo[winnerVote].numberOfVotes) {
+        winnerVote = vote;
+        winnerId = _blockHash;
+        winnerEpoch = _epoch;
+        winnerDrMerkleRoot = _drMerkleRoot;
+        winnerTallyMerkleRoot = _tallyMerkleRoot;
+        //epochFinalizedBlock[_epoch].finalized = false;
+    }
+
+    }
+
+    return bytes32(vote);
+
+  }
+
+  /// @dev Post new block into the block relay
+  /// @param _vote Vote created when the block was proposed
+  /// @param _blockHash Hash of the block headerPost
+  /// @param _epoch Witnet epoch to which the block belongs to
+  /// @param _drMerkleRoot Merkle root belonging to the data requests
+  /// @param _tallyMerkleRoot Merkle root belonging to the tallies
+  /// @param _previousVote Hash of block's hashes proposed in the previous epoch
+  function postNewBlock(
+    uint256 _vote,
+    uint256 _blockHash,
+    uint256 _epoch,
+    uint256 _drMerkleRoot,
+    uint256 _tallyMerkleRoot,
+    uint256 _previousVote)
+    private
+    blockDoesNotExist(_blockHash)
+  {
+    // Map the epoch to the vote's Hashes
+    epochFinalizedBlock[_epoch] = voteInfo[_vote].voteHashes.blockHash;
+    blocks[_blockHash].drHashMerkleRoot = _drMerkleRoot;
+    blocks[_blockHash].tallyHashMerkleRoot = _tallyMerkleRoot;
+    blocks[_blockHash].previousVote = _previousVote;
+
+    // Select previous vote and corresponding epoch
+    uint256 previousVote = blocks[epochFinalizedBlock[_epoch]].previousVote;
+    uint256 epoch = voteInfo[previousVote].voteHashes.epoch;
+
+    // Finalize the previous votes when the corresponding epochs are bigger than the last finalized epoch
+    while (epoch > lastBlock.epoch) {
+      uint256 previousBlockHash = voteInfo[previousVote].voteHashes.blockHash;
+      // Finalize the epoch
+      epochFinalizedBlock[epoch] = previousBlockHash;
+      // Map the block hash to its hashes
+      blocks[previousBlockHash].drHashMerkleRoot = voteInfo[previousVote].voteHashes.drMerkleRoot;
+      blocks[previousBlockHash].tallyHashMerkleRoot = voteInfo[previousVote].voteHashes.tallyMerkleRoot;
+      blocks[previousBlockHash].previousVote = voteInfo[previousVote].voteHashes.previousVote;
+      // Update previousVote and epoch
+      previousVote = blocks[previousBlockHash].previousVote;
+      epoch = voteInfo[previousVote].voteHashes.epoch;
+    }
+    if (epoch == lastBlock.epoch) {
```

Comment:
Check that `epoch` cannot be less than `lastBlock.epoch` with an assert. For example:

```solidity
    assert(epoch == lastBlock.epoch && voteInfo[previousVote].voteHashes.blockHash != lastBlock.blockHash);
```

Reply:
double check it in case I'm wrong :)